src/ucs/datastruct/hlist.h

```diff
@@ -229,7 +237,9 @@ ucs_hlist_extract_head(ucs_hlist_head_t *head)
  * @param _member  List element inside the containing structure.
  */
 #define ucs_list_extract_head_elem(_head, _type, _member) \
-    ucs_container_of(ucs_hlist_extract_head(_head), _type, _member)
+    (!ucs_hlist_is_empty(_head) ? \
+     ucs_container_of(ucs_hlist_extract_head(_head), \
```

Comment:
why needed? IMO it adds extra branch
BTW if it's only optimization, should have been another PR

Reply:
this fixes cppcheck warning about `NULL` subtraction in 255 line

Reply:
w/o this fix can't pass CI

Reply:
re-implemented in the same way as it done in queue implementation